core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,7 +64,23 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // Evaluate partitions in parallel. Partitions of each rdd will be cached by the `partitions`
+  // val in `RDD`.
+  private[spark] lazy val parallelPartitionEval: Boolean = {
```

Comment:
Why would this break in MR? A typical MR job will create one instance of the InputFormat and call getSplits. File listing within that can be done in parallel, but I'm not aware of other cases where frameworks like Hive or Spark call getSplits in parallel. If that's common outside of Spark, then you're right that this is broken elsewhere and we don't need the safety valve.